### PR TITLE
Add hosts and hostname file

### PR DESCRIPTION
**Pull request by ClawSeven**
- Support hosts and hostname file in enclave
- Refine copied file struct
- Modify uname hard code nodename
- Add sethostname, setdomainname syscalls

#604

### CODE REVIEW

**Comment by ClawSeven** [comment]
The parser for hostname and hosts is decoupled with this PR, it will be added in the subsequent PR.

File: src/pal/src/pal_api.c [hunk 1]
```diff
@@ -104,11 +104,24 @@ int occlum_pal_init(const struct occlum_pal_attr *attr) {
...
+    };
+
+    const struct host_file_buffer *file_buffer_ptr = &file_buffer;
+
     sgx_status_t ecall_status = occlum_ecall_init(eid, &ecall_ret, attr->log_level,
-                                resolved_path, resolv_conf_ptr);
-    free((void *)resolv_conf_ptr);
-    resolv_conf_ptr = NULL;
+                                resolved_path, file_buffer_ptr);
+
```
Comment (liqinggd) [inline src/pal/src/pal_api.c:117]: Why not implement a function to implement the free of the `host_file_buffer`?

File: src/pal/src/pal_load_file.c [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include "pal_log.h"
+
+char *pal_load_file(const char *filename) {
```
Comment (liqinggd) [inline src/pal/src/pal_load_file.c:5]: I think the name should be `pal_load_file_to_string`

File: src/pal/src/pal_load_file.c [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    fseek(fp, 0, SEEK_END);
+    long fsize = ftell(fp);
+    fseek(fp, 0, SEEK_SET);
+    char *file_buffer = malloc(fsize + 1);
+    if (file_buffer == NULL) {
+        PAL_WARN("Warning: Failed to malloc buffer for file: %s", filename);
+        return NULL;
+    }
+    fread(file_buffer, 1, fsize, fp);
+    file_buffer[fsize] = 0;
```
Comment (liqinggd) [inline src/pal/src/pal_load_file.c:21]: Use '\0' is more readable than 0

File: src/pal/include/occlum_pal_api.h [hunk 4]
```diff
@@ -93,6 +93,13 @@ struct occlum_pal_create_process_args {
     int *pid;
 };
 
+// todo: more detailed description
+struct host_file_buffer {
+    const char *resolv_conf_ptr;
```
Comment (liqinggd) [inline src/pal/include/occlum_pal_api.h:98]: I think `_ptr` is not accurate, maybe we can use `resolv_conf_buf` `hosts_buf` `hostname_buf`

File: src/pal/src/pal_api.c (hunk 1 above)
Comment (ClawSeven) [reply src/pal/src/pal_api.c:117]: Done

File: src/pal/src/pal_load_file.c (hunk 2 above)
Comment (ClawSeven) [reply src/pal/src/pal_load_file.c]: Done

File: src/pal/src/pal_load_file.c (hunk 3 above)
Comment (ClawSeven) [reply src/pal/src/pal_load_file.c]: Done, nice advice!

File: src/pal/include/occlum_pal_api.h (hunk 4 above)
Comment (ClawSeven) [reply src/pal/include/occlum_pal_api.h]: Done

File: src/libos/src/misc/uname.rs [hunk 5]
```diff
@@ -48,3 +50,43 @@ fn copy_from_cstr_to_u8_array(src: &CStr, dst: &mut [u8]) {
     dst[..len].copy_from_slice(&src[..len]);
     dst[len] = 0;
 }
+
+fn obtain_nodename(dst: &mut [u8]) {
```
Comment (guzongmin) [inline src/libos/src/misc/uname.rs:54]: Is it possible that we get store the hostname in parse_host_file(HostFile::HOSTNAME, hostname_ptr)? The current behavior is: get the hostname from host → parse the file and then save it as a file in SEFS → read the file to get the hostname. Is it too complex?

File: src/libos/include/edl/occlum_edl_types.h [hunk 6]
```diff
@@ -29,6 +29,13 @@ typedef struct itimerspec{
     struct _timespec it_value;
 } itimerspec_t;
 
+// todo: more detailed description
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/include/edl/occlum_edl_types.h]: Done

File: src/libos/src/misc/uname.rs (hunk 5 above)
Comment (ClawSeven) [reply src/libos/src/misc/uname.rs]: Done

File: src/libos/src/misc/rlimit.rs [hunk 7]
```diff
@@ -33,7 +33,7 @@ impl Default for ResourceLimits {
         let address_space = rlimit_t::new(cfg_heap_size + cfg_stack_size + cfg_mmap_size);
 
         // Set init open files limit to 1024 which is default value for Ubuntu
-        let open_files = rlimit_t::new(1024);
+        let open_files = rlimit_t::new(2048);
```
Comment (guzongmin) [inline src/libos/src/misc/rlimit.rs:36]: Does any document support that change the default value from 1024 to 2048?

File: src/libos/src/fs/file_table.rs [hunk 8]
```diff
@@ -65,9 +65,20 @@ impl FileTable {
         Ok(min_free_fd)
     }
 
-    pub fn put(&mut self, file: FileRef, close_on_spawn: bool) -> FileDesc {
+    pub fn put(&mut self, file: FileRef, close_on_spawn: bool) -> Result<FileDesc> {
         let mut table = &mut self.table;
 
+        let soft_rlimit_nofile = current!()
```
Comment (guzongmin) [inline src/libos/src/fs/file_table.rs:71]: I find that the "fn dup" at line 28 and "fn put_at" at line 100 need a similar check.

**Review by guzongmin — APPROVED** [review]
LGTM

**Pull request merged**